### PR TITLE
Vectorize Recency Sampler Using Circular Buffers

### LINKED ISSUE #61: Vectorize the Recency Sampler Hook

### PR DESCRIPTION
**Pull request by Jacob-Chmura**
# Purpose
The purpose of this PR is to make an implementation change in our recency sampler hook to speed up it's execution.

Specifically, I moved away from python queues and wrote a fully torch-native implementation using circular buffers, which can be stored on GPU.

### Outcome (**independent of sparse dedup hook improvements** #60)
End to end latency reduction by ~81%

#### Case (`perf/dict_idx`)
Screenshot from 2025-06-24 16-17-56

#### Control (`perf/recency`)
Screenshot from 2025-06-25 08-26-21

## Key Changes
- TODO

## Relevant Prs
Close #61

### CODE REVIEW

File: examples/linkproppred/tgat.py [hunk 1]
```diff
@@ -223,8 +225,11 @@ def _init_hooks(dg: DGraph, sampling_type: str) -> List[DGHook]:
 test_metrics = MetricCollection(metrics, prefix='Test')
 
 for epoch in range(1, args.epochs + 1):
+    from tgm.util.perf import Profiling
+
     start_time = time.perf_counter()
-    loss = train(train_loader, model, opt)
+    with Profiling(filename='foo'):
+        loss = train(train_loader, model, opt)
```
Comment (Jacob-Chmura) [inline examples/linkproppred/tgat.py]: TODO; remove profilling

File: examples/linkproppred/tgat.py [hunk 2]
```diff
@@ -188,7 +186,11 @@ def _init_hooks(dg: DGraph, sampling_type: str) -> List[DGHook]:
     if sampling_type == 'uniform':
         nbr_hook = NeighborSamplerHook(num_nbrs=args.n_nbrs)
     elif sampling_type == 'recency':
-        nbr_hook = RecencyNeighborHook(num_nbrs=args.n_nbrs, num_nodes=dg.num_nodes)
+        nbr_hook = RecencyNeighborHook(
+            num_nbrs=args.n_nbrs,
+            num_nodes=dg.num_nodes,
+            edge_feats_dim=dg.edge_feats_dim,
+        )
```
Comment (Jacob-Chmura) [inline examples/linkproppred/tgat.py:187]: TODO: Update other examples with the api change (pass forward edge feats dim)

File: tgm/hooks.py [hunk 3]
```diff
@@ -272,20 +248,32 @@ class RecencyNeighborHook:
...
-        for node in range(num_nodes):
-            self._nbrs[node] = [deque(maxlen=max(num_nbrs)) for _ in range(3)]
+        self._max_nbrs = max(num_nbrs)
+
+        self._nbr_nids = torch.full((num_nodes, self._max_nbrs), -1, dtype=torch.long)
+        self._nbr_times = torch.zeros((num_nodes, self._max_nbrs), dtype=torch.long)
+        self._nbr_feats = torch.zeros((num_nodes, self._max_nbrs, edge_feats_dim))
+        self._nbr_mask = torch.zeros((num_nodes, self._max_nbrs), dtype=torch.bool)
+
+        # Circular buffer ptr
```
Comment (shenyangHuang) [inline tgm/hooks.py:308]: this is doing all the work it seems like?

File: tgm/hooks.py [hunk 4]
```diff
@@ -294,63 +282,73 @@ def num_nbrs(self) -> List[int]:
...
+            s, d, t = int(src[i].item()), int(dst[i].item()), int(time[i].item())
+            for u, v in [(s, d), (d, s)]:
+                ptr = int(self._nbr_ptr[u].item())
+                self._nbr_nids[u, ptr] = v
+                self._nbr_times[u, ptr] = t
+                self._nbr_mask[u, ptr] = True
+                if batch.edge_feats is not None:
+                    self._nbr_feats[u, ptr] = batch.edge_feats[i]
+
+                self._nbr_ptr[u] = (ptr + 1) % self._max_nbrs
```
Comment (shenyangHuang) [inline tgm/hooks.py]: circle and overwrite

File: tgm/hooks.py [hunk 5]
```diff
@@ -293,7 +303,7 @@ def __call__(self, dg: DGraph, batch: DGBatch) -> DGBatch:
                     # leakage, making the prediction easier than it should be.
                     fake_times = torch.randint(
                         int(batch.time.min().item()),
-                        int(batch.time.max().item()),
+                        int(batch.time.max().item()) + 1,
```
Comment (shenyangHuang) [inline tgm/hooks.py:340]: this is to be inclusive of the maxtime right

File: tgm/hooks.py [hunk 6]
```diff
@@ -293,7 +303,7 @@ def __call__(self, dg: DGraph, batch: DGBatch) -> DGBatch:
                     # leakage, making the prediction easier than it should be.
```
Comment (shenyangHuang) [inline tgm/hooks.py:337]: what does this mean? should we remove the leakage comment?

File: tgm/hooks.py (hunk 5 above)
Comment (shenyangHuang) [reply tgm/hooks.py:340]: maybe add a comment for us to remember in the future 
timestamps for negatives destinations in a batch is sampled from `[batch_min_time, batch_max_time]` inclusive.

File: tgm/hooks.py [hunk 7]
```diff
@@ -305,42 +315,50 @@ def __call__(self, dg: DGraph, batch: DGBatch) -> DGBatch:
...
+            batch.nbr_feats.append(nbr_feats)  # type: ignore
+            batch.nbr_mask.append(nbr_mask)  # type: ignore
 
         self._update(batch)
         return batch
 
+    def _get_recency_indices(self, node_ids: torch.Tensor, k: int) -> torch.Tensor:
+        ptr = self._nbr_ptr[node_ids].unsqueeze(1)
+        offsets = torch.arange(k, device=node_ids.device).unsqueeze(0)
+        indices = (ptr - 1 - offsets) % self._max_nbrs
```
Comment (shenyangHuang) [inline tgm/hooks.py:373]: looks correct, though with pointer and everything, should make sure it is working as intended. What happens when a node has no neighbors whatsoever?

File: tgm/hooks.py (hunk 6 above)
Comment (Jacob-Chmura) [reply tgm/hooks.py:337]: This explains why we chose our heuristic for assigning timestamps on negative edges. We don't do random timestamps because this would give an implicit signal to the model about which edges are positive and negative.

File: tgm/hooks.py (hunk 5 above)
Comment (Jacob-Chmura) [reply tgm/hooks.py:340]: Yeah will do
